gadget/update.go

```diff
@@ -904,6 +905,247 @@ func buildNewVolumeToDeviceMapping(old GadgetData, laidOutVols map[string]*LaidO
+			// all the math to find the correct start offset for structures that
+			// don't explicitly list their offset in the gadget.yaml (and thus
+			// would have a offset of 0 using the volStruct.Offset)
+			structStartOffset := laidOutVol.LaidOutStructure[volYamlIndex].StartOffset
+
+			if volStruct.HasFilesystem() {
+				// Here we know what disk is associated with this volume, so we
+				// just need to find what partition is associated with this
+				// structure to find it's root mount points. On GPT since
+				// partition labels/names are unique in the partition table, we
+				// could do a lookup by matching partition label, but this won't
+				// work on MBR which doesn't have such a concept, so instead we
+				// use the start offset to locate which disk partition this
+				// structure is equal to.
+
+				partitions, err := disk.Partitions()
+				if err != nil {
+					return nil, err
+				}
+
+				var foundP disks.Partition
+				found := false
+				for _, p := range partitions {
+					if p.StartInBytes == uint64(structStartOffset) {
+						foundP = p
+						found = true
+						break
+					}
+				}
+				if !found {
+					dieErr := fmt.Errorf("cannot locate structure %d on volume %s: no matching start offset", volYamlIndex, volName)
+					if fatal := maybeDieOnMappingError(dieErr); fatal {
+						return nil, dieErr
+					} else {
+						return nil, errSkipUpdateProceedRefresh
+					}
+				}
+
+				// if this structure is an encrypted one, then we can't just
+				// get the root mount points for the device node, we would need
+				// to find the decrypted mapper device for the encrypted device
+				// node and then find the root mount point of the mapper device
+				if _, ok := diskDeviceTraits.StructureEncryption[volStruct.Name]; ok {
+					logger.Noticef("gadget asset update for assets on encrypted partition %s unsupported", volStruct.Name)
+
+					// setting this structure as having an empty location will
+					// mean when an update to this structure is actually
+					// performed it will fail, but we won't fail updates to
+					// other structures - it is treated like an unmounted
+					// partition
+					volumeStructureToLocation[volName][volYamlIndex] = StructureLocation{}
+					continue
+				}
+
+				// otherwise normal unencrypted filesystem, find the rw mount
+				// points
+				mountpts, err := disks.MountPointsForPartitionRoot(foundP, map[string]string{"rw": ""})
+				if err != nil {
+					dieErr := fmt.Errorf("cannot locate structure %d on volume %s: error searching for root mount points: %v", volYamlIndex, volName, err)
+					if fatal := maybeDieOnMappingError(dieErr); fatal {
+						return nil, dieErr
+					} else {
+						return nil, errSkipUpdateProceedRefresh
+					}
+				}
+				var mountpt string
+				if len(mountpts) == 0 {
+					// this filesystem is not already mounted, we probably
+					// should mount it in order to proceed with the update?
+
+					// TODO: do something better here?
```

Comment:
do we have use cases like this? or all the special partitions raw?

Reply:
The use case here is not when the partition is raw, but rather when there is a partition with a filesystem, but it is not automatically mounted by anything in Linux (i.e. imagine a FAT partition that is used by some sort of co-processor), and the content on that partition is updated via a gadget asset update.

We have the same problem with the current code today too, where if it's not mounted then it doesn't get updated, when I asked you about this you said it was fine to leave with a TODO about what to do later since it's not a change from current behavior

See `newMountedFilesystemUpdater`, if the partition is not mounted we fail with the error message `cannot find mount location of structure foo: ...`